src/System.Runtime.WindowsRuntime/src/System/IO/WindowsRuntimeStorageExtensions.cs

```diff
@@ -212,9 +174,146 @@ public static Task<Stream> OpenStreamForReadAsync(this IStorageFolder rootDirect
+            HANDLE_ACCESS_OPTIONS accessOptions = FileAccessToHandleAccessOptions(access);
+            HANDLE_SHARING_OPTIONS sharingOptions = FileShareToHandleSharingOptions(share);
+            HANDLE_OPTIONS handleOptions = FileOptionsToHandleOptions(options);
+
+            IStorageFolderHandleAccess handleAccess = ((object)rootDirectory) as IStorageFolderHandleAccess;
+
+            if (handleAccess == null)
+                return null;
+
+            SafeFileHandle handle;
+
+            int result = handleAccess.Create(
+                relativePath,
+                creationOptions,
+                accessOptions,
+                sharingOptions,
+                handleOptions,
+                IntPtr.Zero,
+                out handle);
+
+            if (result != 0)
+                throw Win32Marshal.GetExceptionForWin32Error(Win32Marshal.TryMakeWin32ErrorCodeFromHR(result), relativePath);
+
+            return handle;
+        }
+
+        private static HANDLE_ACCESS_OPTIONS FileAccessToHandleAccessOptions(FileAccess access)
+        {
+            switch (access)
+            {
+                case FileAccess.ReadWrite:
+                    return HANDLE_ACCESS_OPTIONS.HAO_READ | HANDLE_ACCESS_OPTIONS.HAO_WRITE;
+                case FileAccess.Read:
+                    return HANDLE_ACCESS_OPTIONS.HAO_READ;
+                case FileAccess.Write:
+                    return HANDLE_ACCESS_OPTIONS.HAO_WRITE;
+            }
+
+            throw new ArgumentOutOfRangeException(nameof(access));
+        }
+
+        private static HANDLE_SHARING_OPTIONS FileShareToHandleSharingOptions(FileShare share)
+        {
+            if ((share & FileShare.Inheritable) != 0)
+                throw new NotSupportedException(SR.NotSupported_Inheritable);
+            if (share < FileShare.None || share > (FileShare.ReadWrite | FileShare.Delete))
+                throw new ArgumentOutOfRangeException(nameof(share));
+
+            HANDLE_SHARING_OPTIONS sharingOptions = HANDLE_SHARING_OPTIONS.HSO_SHARE_NONE;
+            if ((share & FileShare.Read) != 0)
+                sharingOptions |= HANDLE_SHARING_OPTIONS.HSO_SHARE_READ;
+            if ((share & FileShare.Write) != 0)
+                sharingOptions |= HANDLE_SHARING_OPTIONS.HSO_SHARE_WRITE;
+            if ((share & FileShare.Delete) != 0)
+                sharingOptions |= HANDLE_SHARING_OPTIONS.HSO_SHARE_DELETE;
+
+            return sharingOptions;
+        }
+
+        private static HANDLE_OPTIONS FileOptionsToHandleOptions(FileOptions options)
+        {
+            if ((options & FileOptions.Encrypted) != 0)
+                throw new NotSupportedException(SR.NotSupported_Encrypted);
+            if (options != FileOptions.None && (options &
+                ~(FileOptions.WriteThrough | FileOptions.Asynchronous | FileOptions.RandomAccess | FileOptions.DeleteOnClose |
+                  FileOptions.SequentialScan | (FileOptions)0x20000000 /* NoBuffering */)) != 0)
```

Comment:
I hate that we've got these valid file options values that aren't defined in FileOptions.

Not really relevant to this PR, just a thought. This isn't the only place we do this kind of thing.